### ISSUE TITLE
no-null should allow, or have an option to allow, the `null == foo` pattern for value-presence-checking

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [sindresorhus/eslint-plugin-unicorn#1507]
A common pattern for checking for both `undefined` and `null` in a single operation is to intentionally use the coercive equality operator with `null` as the comparee:

```js
if (foo == null)
   // ...
```

At the moment, this triggers the otherwise-wonderful `no-null` lint.

In the codebases I've worked on, this *specific* pattern is the *only* place that either:

1. coerceive equality, or
2. `null`

... is allowed. At least as far as I'm aware, that's a common situation (those two being allowed in this specific pattern.)

Personally, despite the modern world with linting and TypeScript, I also still prefer yoda-comparisons; so I'd appreciate if this also allows (or the option also allows, if we add an option for this) the reversed version:

```js
if (null == foo)
   // ...
```

I hope this Issue was helpful, and not just a waste of everyone's time! 😅

**Comment by ELLIOTTCABLE** [comment]
(Also, the Issue Templates don't currently have 'feature-request-but-for-existing-rule'; and out of 'create an entirely new rule' and 'small bug in existing rule', I suspected this Issue was closest to the latter — but apologies if I mis-categorized!)

**Comment by sindresorhus** [comment]
The whole point of the rule is to encourage using `undefined` instead of `null`. Allowing `foo == null` will let `null` be used where `undefined` could be used instead. I cannot tell how you're using `foo == null`, but it should probably be `foo === undefined` instead, or alternatively `foo === undefined || foo === null` (explicitness is better for readability).

**Comment by ELLIOTTCABLE** [comment]
1. "Where would you want it" - you can excise null from *your own* codebase (great!), but you should still be checking for it at API entrypoints. Code defensively!
2. Well, yes, although that's a matter of taste — two checks has a vanishingly minimal perf impact, of course; but mostly I'm of the camp that thinks that's way, way too much text for a feature that's explicitly built into the language (i.e. the explicit "checking for null and undefined, but no other values" expressed by that loose-comparison.)

I won't be offended if this isn't to y'all's tastes and you decide to close this, though. ¯\_(ツ)_/¯

**Comment by fregante** [comment]
If you want to use that pattern you'd have to set this:

```js
"eqeqeq": ["error", "smart"],
"unicorn/no-null": "off",
```

but of course that means that `null` can still be used everywhere else as well.

**Comment by sindresorhus** [comment]
Can you not use `if (foo != undefined) {}` instead? It also matches `null`.

**Comment by sindresorhus** [comment]
> Saying 'don't use null == value because it's less explicit' feels a bit like saying 'don't use foo?.bar because if (typeof foo !== 'undefined' && foo !== null) { foo.bar } is more explicit'. That's why we have syntactic sugar for common patterns! :P

Well.. I was arguing that optional chaining should only include `undefined` back when it was conceived, but it was determined that it was too unrealistic for the ecosystem at the time. Also, default function arguments, destructuring defaults, and many other things only work with undefined.

**Comment by ELLIOTTCABLE** [comment]
I think I'm in a similar place re: "the ecosystem at the time", except I'm probably waaaaay to conservative in my estimation of "at the time." I'm kinda a JavaScripter who's coding defensively at all times as if it's 2009. 🤣

The *completely* ideal solution, tbh, would be "only allow `null` to appear 1. in a defensively-checking-position when 2. the defensive-check appears in an 'API boundary' function", i.e. one `export`ed from the file … but that's probably a bit too much to ask.

Anyway, since this isn't a bug, and by their very nature, linting configs are very personal/subjective things … I think I'll go with the "principal of enthusiastic consent" and close this Issue for now. <3
